libraries/botbuilder-dialogs-adaptive-testing/src/dialogInspector.ts

```diff
@@ -0,0 +1,175 @@
+/**
+ * @module botbuilder-dialogs-adaptive-testing
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { BotStateSet, ConversationState, TurnContext, TurnContextStateCollection, UserState } from 'botbuilder-core';
+import {
+    Dialog,
+    DialogContext,
+    DialogSet,
+    DialogState,
+    DialogStateManager,
+    DialogStateManagerConfiguration,
+} from 'botbuilder-dialogs';
+import { LanguageGeneratorManager, languageGeneratorManagerKey } from 'botbuilder-dialogs-adaptive';
+import { ResourceExplorer } from 'botbuilder-dialogs-declarative';
+import { getActiveDialogContext } from 'botbuilder-dialogs/lib/dialogHelper';
+
+const CONVERSATION_STATE = 'ConversationState';
+const USER_STATE = 'UserState';
+
+export type DialogContextInspector = (dc: DialogContext) => void;
+
+/**
+ * Class for inspecting current dialog context.
+ */
+export class DialogInspector {
+    private _rootDialogId: string;
+    private readonly _dialogStateProperty: string;
+
+    /**
+     * Initializes a new instance of the `DialogInspector` class.
+     *
+     * @param {Dialog} rootDialog Root dialog to use.
+     * @param {ResourceExplorer} resourceExplorer Resource explorer for expression access to .lg templates.
+     * @param {string} dialogStateProperty Alternate name for the dialogState property. (Default is "DialogState").
+     */
+    public constructor(rootDialog?: Dialog, resourceExplorer?: ResourceExplorer, dialogStateProperty?: string) {
+        if (!resourceExplorer) {
+            // Add language generator for function access
+            this.initialTurnState.set(languageGeneratorManagerKey, new LanguageGeneratorManager(resourceExplorer));
+        }
+
+        if (rootDialog) {
+            this.rootDialog = rootDialog;
+        }
+
+        this._dialogStateProperty = dialogStateProperty ?? 'DialogState';
+    }
+
+    /**
+     * Gets or sets the ConversationState.
+     */
+    public conversationState: ConversationState;
+
+    /**
+     * Gets or sets the UserState.
+     */
+    public userState: UserState;
+
+    /**
+     * Gets initialTurnState collection to copy into the turnState on every turn.
+     */
+    public readonly initialTurnState = new TurnContextStateCollection();
+
+    /**
+     * Gets root dialog to use to start conversation.
+     *
+     * @returns {Dialog} The root dialog.
+     */
+    public get rootDialog(): Dialog {
```

Comment:
If possible, add `| undefined` to the return type signature. In strict mode, `undefined` does not satisfy any types aside from `undefined`.